# Tarea - Programación lineal
## Úrsula Vargas López  740388
## 20/02/2023

### Diseño de la Dieta Óptima

Se quiere producir comida para gatos de la manera más barata, no obstante se debe también asegurar que se cumplan los datos requeridos de analisis nutricional. Por lo que se quiere variar la cantidad de cada ingrediente para cumplir con los estandares nutricionales. Los requisitos que se tienen es que en $100$ gramos, se deben tener **por lo menos** $8$ gramos de proteína y $6$ gramos de grasa. Así mismo, no se debe tener más de $2$ gramos de fibra y $0.4$ gramos de sal.  

Los datos nutricionales se pueden obtener de la siguiente tabla:

Ingrediente|Proteína|Grasa|Fibra|Sal
:----|----|----|----|----
Pollo|  10.0%|08.0%|00.1%|00.2%
Carne|  20.0%|10.0%|00.5%|00.5%
Cordero|15.0%|11.0%|00.5%|00.7%
Arroz|  00.0%|01.0%|10.0%|00.2%
Trigo|  04.0%|01.0%|15.0%|00.8%
Gel|    00.0%|00.0%|00.0%|00.0%

Los costos de cada producto son:

Ingrediente|Costo por gramo
:----|----
Pollo|$\$$0.013
Carne|$\$$0.008
Cordero|$\$$0.010
Arroz|$\$$0.002
Trigo|$\$$0.005
Gel|$\$$0.001    

Lo que se busca optimizar en este caso es la cantidad de productos que se debe utilizar en la comida de gato, minimizando el costo total. Para simplificar la notación use las siguientes variables: 

+ $x_1:$ Gramos de pollo  
+ $x_2:$ Gramos de carne  
+ $x_3:$ Gramos de cordero  
+ $x_4:$ Gramos de arroz  
+ $x_5:$ Gramos de trigo  
+ $x_6:$ Gramos de gel

La tarea consiste en plantear el problemade programación lineal que permita satisfacer las necesidades alimenticias del gato al tiempo que minimice el costo total y resolverlo con `linprog`.

### PLANTEAMIENTO

1.-Minimizar costo total

2.- Variables $X1,X2,X3,X4,X5,X6$

3.-Función de costo

$=0.013x1+0.008x2+0.010x3+0.002x4+0.005x5+0.001x6$

4.-Restricciones

$-0.1x1-0.2x2-0.15x3-0x4-0.04x5-0x6<=-8$

$-0.08x1-0.1x2-0.11x3-0.01x4-0.01x5-0x6<=-6$

$0.001+0.005+0.005+0.10+0.15+0<=2$

$0.002+0.007+0.007+0.002+0.008+0<=0.4$


In [4]:
import scipy.optimize as opt
import numpy as np

c=np.array([
    [0.013,
     0.008,
     0.010,
     0.002,
     0.005,
     0.001]
])

Aub= np.array([
    [-0.1,-0.2,-0.15,0,-0.04,0],
    [-0.08,-0.1,-0.11,-0.01,-0.01,0],
    [0.001,0.005,0.005,0.10,0.15,0],
    [0.002,0.007,0.007,0.002,0.008,0]
])

bub=np.array([
    [-8,
     -6,
     2,
     0.40]
])

Aeq=np.array([
    [1,
     1,
     1,
     1,
     1,
     1]
])

beq=np.array([
    [100]
])

opt.linprog (c,Aub,bub,Aeq,beq)
     

           con: array([0.])
 crossover_nit: 0
         eqlin:  marginals: array([0.001])
  residual: array([0.])
           fun: 0.5555555555555554
       ineqlin:  marginals: array([-0.        , -0.19444444, -0.        , -1.77777778])
  residual: array([3.66666667, 0.        , 1.71666667, 0.        ])
         lower:  marginals: array([0.00000000e+00, 0.00000000e+00, 5.55555556e-05, 2.61111111e-03,
       1.62777778e-02, 0.00000000e+00])
  residual: array([ 5.55555556, 55.55555556,  0.        ,  0.        ,  0.        ,
       38.88888889])
       message: 'Optimization terminated successfully. (HiGHS Status 7: Optimal)'
           nit: 3
         slack: array([3.66666667, 0.        , 1.71666667, 0.        ])
        status: 0
       success: True
         upper:  marginals: array([0., 0., 0., 0., 0., 0.])
  residual: array([inf, inf, inf, inf, inf, inf])
             x: array([ 5.55555556, 55.55555556,  0.        ,  0.        ,  0.        ,
       38.88888889])

## RESPUESTA

#### Para minimizar el costo total y satisfacer las necesidades alimenticias del gato se necesitan 5.55gr de pollo, 55.55gr de carne y 38.88gr de gel.
#### Con un costo de producción de $0.55 por pieza(bolsa)

### 2.- Problema de distribución de energía eléctrica

La Comisión Federal de Electricidad **(CFE)** dispone de cuatro plantas de generación para satisfacer la demanda diaria eléctrica en cuatro ciudades, Guadalajara, León y Morelia. Las plantas $1$, $2$ y $3$ pueden satisfacer $80$, $40$ y $60$ millones de KW al día respectivamente. Las necesidades de las ciudades de Guadalajara, León y Morelia son de $70$, $40$ y $70$ millones de Kw al día respectivamente. 


Los costos asociados al envío de suministro energético por cada millón de Kw entre cada planta y cada ciudad son los registrados en la siguiente tabla. 

-|Guadalajara|León|Morelia
:----|----|----|----
Planta 1|5|2|7
Planta 2|3|6|6
Planta 3|6|1|2

Y por último, las restricciones del problema, van a estar dadas por las capacidades de oferta y demanda de cada planta (en millones de KW) y cada ciudad.

Para simplificar la notación use las siguientes variables:

+ $x_1$: Kw (en millones) distribuidos de la Planta 1 a Guadalajara
+ $x_2$: Kw (en millones) distribuidos de la Planta 1 a León
+ $x_3$: Kw (en millones) distribuidos de la Planta 1 a Morelia
+ $x_4$: Kw (en millones) distribuidos de la Planta 2 a Guadalajara
+ $x_5$: Kw (en millones) distribuidos de la Planta 2 a León
+ $x_6$: Kw (en millones) distribuidos de la Planta 2 a Morelia
+ $x_7$: Kw (en millones) distribuidos de la Planta 3 a Guadalajara
+ $x_8$: Kw (en millones) distribuidos de la Planta 3 a León
+ $x_9$: Kw (en millones) distribuidos de la Planta 3 a Morelia

La tarea consiste en plantear el problema de programación lineal que permita satisfacer las necesidades de todas las ciudades al tiempo que minimice los costos asociados a la distribución y resolverlo con `linprog`.

## PLANTEAMINETO

1.-Minimizar costos de distribución

2.- Variables $X1,X2,X3,X4,X5,X6,X7,X8,X9$

3.-Función costo

$=5x1+2x2+7x3+3x4+6x5+6x6+6x7+x8+2x9$

4.-Restricciones

PLANTAS

$x1+x2+x3<=80$

$x4+x5+x6<=40$

$x7+x8+x9<=60$


CIUDADES

$x1+x4+x7=70$

$x2+x5+x8=40$

$x3+x6+x9=70$



In [1]:
import scipy.optimize as opt
import numpy as np

c=np.array ([
    [5,
     2,
     7,
     3,
     6,
     6,
     6,
     1,
     2]
     ])
    
Aub=np.array ([
    [1,1,1,0,0,0,0,0,0],
    [0,0,0,1,1,1,0,0,0],
    [0,0,0,0,0,0,1,1,1]
    ])

bub=np.array([
    [80,
     40,
     60]
])

Aeq=np.array([
    [1,0,0,1,0,0,1,0,0],
    [0,1,0,0,1,0,0,1,0],
    [0,0,1,0,0,1,0,0,1]
])

beq=np.array([
    [70,
     40,
     70]
])

opt.linprog(c,Aub,bub,Aeq,beq)
    
    
    

           con: array([0., 0., 0.])
 crossover_nit: 0
         eqlin:  marginals: array([5., 2., 7.])
  residual: array([0., 0., 0.])
           fun: 540.0
       ineqlin:  marginals: array([-0., -2., -5.])
  residual: array([0., 0., 0.])
         lower:  marginals: array([0., 0., 0., 0., 6., 1., 6., 4., 0.])
  residual: array([30., 40., 10., 40.,  0.,  0.,  0.,  0., 60.])
       message: 'Optimization terminated successfully. (HiGHS Status 7: Optimal)'
           nit: 6
         slack: array([0., 0., 0.])
        status: 0
       success: True
         upper:  marginals: array([0., 0., 0., 0., 0., 0., 0., 0., 0.])
  residual: array([inf, inf, inf, inf, inf, inf, inf, inf, inf])
             x: array([30., 40., 10., 40.,  0.,  0.,  0.,  0., 60.])

# RESPUESTA

### Para minimizar el costo de distribución se deben enviar:
##### 30  Kw (en millones) distribuidos de la Planta 1 a Guadalajara
##### 40  Kw (en millones) distribuidos de la Planta 1 a León
##### 10  Kw (en millones) distribuidos de la Planta 1 a Morelia
##### 40  Kw (en millones) distribuidos de la Planta 2 a Guadalajara
##### 60  Kw (en millones) distribuidos de la Planta 3 a Morelia

##### Con un costo total de distribución de $540